### PR TITLE
Adding new command to update PHP version for code studio.

### PR DESCRIPTION
**Pull request by joshirohit100**
Issue - https://github.com/acquia/cli/issues/1297
Screenshot 2023-01-04 at 6 31 42 PM

### CODE REVIEW

**Comment by joshirohit100** [comment]
Added initial stab for review. If look good, will add test case as well

**Comment by anavarre** [comment]
I'm not sure I fully understand the value of this command but what I know is it should be `codestudio:php-version` to have the same command name as `ide:php-version` per https://github.com/acquia/cli/blob/83dfb30525e7f7bea326c3cc53b00566f2474d68/src/Command/Ide/IdePhpVersionCommand.php#L21

Also, we should likely extract some paradigms from this IDE file, such as https://github.com/acquia/cli/blob/83dfb30525e7f7bea326c3cc53b00566f2474d68/src/Command/Ide/IdePhpVersionCommand.php#L87

What are your thoughts @danepowell ?

File: src/Command/CodeStudio/CodeStudioChangePHPVersionCommand.php [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    // If no project found matching with attached Acquia CLoud App id.
+    if (empty($project_id)) {
+      $this->io->error('There are no Acquia Cloud projects available in Code Studio.');
+      return 1;
+    }
+
+    try {
+      // If PHP version is not set in variables.
+      if (!$php_version_already_set) {
+        if ($php_version === '8.0') {
```
Comment (danepowell) [inline src/Command/CodeStudio/CodeStudioChangePHPVersionCommand.php:81]: Will this _always_ be the default? I doubt it. I'd rather not hardcode this message and have it become stale when the default version changes.

File: src/Command/CodeStudio/CodeStudioChangePHPVersionCommand.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+  /**
+   * @param \Symfony\Component\Console\Input\InputInterface $input
+   * @param \Symfony\Component\Console\Output\OutputInterface $output
+   *
+   * @return int
+   * @throws \Exception
+   */
+  protected function execute(InputInterface $input, OutputInterface $output): int {
+    $php_version = $input->getArgument('php-version');
```
Comment (danepowell) [inline src/Command/CodeStudio/CodeStudioChangePHPVersionCommand.php:37]: Validate the PHP version. Use https://github.com/acquia/cli/blob/83dfb30525e7f7bea326c3cc53b00566f2474d68/src/Command/Ide/IdePhpVersionCommand.php#L87 . Move that method out of the IdeCommandTrait to somewhere more common like the command base, and pull the last line that's specific to IDEs.

**Review by danepowell — CHANGES_REQUESTED** [review]
This is a great start, but it will also need unit tests.

File: src/Command/CodeStudio/CodeStudioChangePHPVersionCommand.php [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+  protected static $defaultName = 'codestudio:update-php-version';
+
+  /**
+   * {inheritdoc}.
+   */
+  protected function configure(): void {
+    $this->setDescription('Change the PHP version in Code Studio')
+      ->addArgument('php-version', InputArgument::REQUIRED, 'The PHP version that needs to configured or updated')
+      ->addOption('gitlab-token', NULL, InputOption::VALUE_REQUIRED, 'The GitLab personal access token that will be used to communicate with the GitLab instance')
```
[the remark this answers is not in the document]
Comment (joshirohit100) [reply src/Command/CodeStudio/CodeStudioChangePHPVersionCommand.php]: Moved to `CodeStudioCommandTrait` with new method `acceptGitlabOptions()`

File: src/Command/CodeStudio/CodeStudioChangePHPVersionCommand.php [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      ->addOption('gitlab-host-name', NULL, InputOption::VALUE_REQUIRED, 'The GitLab hostname.');
```
[the remark this answers is not in the document]
Comment (joshirohit100) [reply src/Command/CodeStudio/CodeStudioChangePHPVersionCommand.php]: Now project id is available as gitlab options moved to common method but not using the projectId in this command. As  fetching the gitlab project based the provided acquia cloud app.

File: src/Command/CodeStudio/CodeStudioChangePHPVersionCommand.php [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    $project_id = 0;
+    $php_version_already_set = FALSE;
+    foreach ($ownedProjects as $project) {
+      // Get all the variables set for the project.
+      $all_project_variables = $this->gitLabClient->projects()->variables($project['id']);
+      if (!empty($all_project_variables)) {
+        $variables = array_column($all_project_variables, 'value', 'key');
+        // Assuming this project is attached with Acquia cloud only when
+        // ACQUIA_APPLICATION_UUID variable is available/configured.
+        if (!empty($variables['ACQUIA_APPLICATION_UUID']) && $variables['ACQUIA_APPLICATION_UUID'] === $acquiaCloudAppId) {
```
[the remark this answers is not in the document]
Comment (joshirohit100) [reply src/Command/CodeStudio/CodeStudioChangePHPVersionCommand.php]: Removed this and not required now.

File: src/Command/CodeStudio/CodeStudioChangePHPVersionCommand.php (hunk 1 above)
Comment (joshirohit100) [reply src/Command/CodeStudio/CodeStudioChangePHPVersionCommand.php]: removed this.

**Comment by joshirohit100** [comment]
Will work on tests now

**Comment by joshirohit100** [comment]
Added the test cases

**Review by joshirohit100 — COMMENTED** [review]
All feedbacks addressed

**Comment by danepowell** [comment]
Thanks so much for the hard work, this is 95% good. Can you please just kill a few of the escaped mutants? You don't need to catch them all, although that would be awesome. Usually this just requires adding more assertions to the test cases.

I know not everyone is familiar with mutation testing, if you need more specific guidance let me know.

File: src/Command/CodeStudio/CodeStudioCommandTrait.php [hunk 6]
```diff
@@ -273,4 +274,16 @@ private function getGitLabProjectDefaults(): array {
     ];
   }
 
+  /**
+   * Add gitlab options to the command.
+   *
+   * @return $this
+   */
+  protected function acceptGitlabOptions(): static {
```
Comment (danepowell) [inline src/Command/CodeStudio/CodeStudioCommandTrait.php:282]: This is an easy mutant to kill, just make this method private if it's not being overridden anywhere.

Comment (joshirohit100) [reply src/Command/CodeStudio/CodeStudioCommandTrait.php]: Fixed!

**Comment by joshirohit100** [comment]
Killed some mutants. There are some still and not sure of how to kill them (Also not sure if need to be killed or not)

**Comment by danepowell** [comment]
Looks good, thanks!

**Pull request merged**